lib/elixir/lib/kernel/typespec.ex

```diff
@@ -478,11 +478,46 @@ defmodule Kernel.Typespec do
   ## Macro callbacks
 
   @doc false
+  def defspec(:callback, expr, caller) do
+    {name, arity} = spec_to_signature(expr)
+
+    name_str = Atom.to_string(name)
+
+    {docs_kind, docs_name, docs_arity} =
+      if match?("MACRO-" <> _, name_str) do
+        "MACRO-" <> macro_name = name_str
+        {:defmacro, String.to_atom(macro_name), arity - 1}
+      else
+        {:def, name, arity}
+      end
+
+    Module.store_typespec(caller.module, :callback, {:callback, expr, caller})
+
+    Kernel.Typespec.ensure_use_behaviour(caller.module)
```

Comment:
Cool, it seemed wrong to do it this way anyway. :)

I will certainly add tests, just wanted to make sure this was going in the right direction before spending any more time. Thanks for the feedback! :+1: